src/main/java/nextstep/subway/line/domain/Sections.java

```diff
@@ -0,0 +1,134 @@
+
+        sections.remove(includingSection);
+        sections.add(addedSection);
+        sections.add(includingSection.subtract(addedSection));
+    }
+
+    private void validateAddedSection(Section addedSection) {
+        List<Long> stationIds = stationIds();
+
+        boolean containsUpStation = stationIds.contains(addedSection.getUpStationId());
+        boolean containsDownStation = stationIds.contains(addedSection.getDownStationId());
+
+        if (containsUpStation && containsDownStation) {
+            throw new CannotAddSectionException("상행역과 하행역이 이미 노선에 모두 등록되어 있으면 구간을 추가할 수 없습니다.");
+        }
+
+        if (!containsUpStation && !containsDownStation) {
+            throw new CannotAddSectionException("상행역과 하행역 둘 중 하나도 포함되어있지 않으면 구간을 추가할 수 없습니다.");
+        }
+
+        sections.stream()
+                .filter(s -> s.startsOrEndsTogether(addedSection) && s.isShorterThanOrEqualWith(addedSection))
+                .findAny()
+                .ifPresent(s -> {
+                    throw new CannotAddSectionException("기존 구간 사이에 추가할 구간의 길이가 기존 구간의 길이보다 크거나 같을 수 없습니다.");
+                });
+    }
+
+    public void removeSection(Long stationId) {
+        if (sections.isEmpty()) {
+            throw new CannotDeleteSectionException("현재 존재하는 구간이 없습니다.");
+        }
+
+        Section lastSection = lastSection();
+        if (!lastSection.matchDownStation(stationId)) {
+            throw new CannotDeleteSectionException("노선의 종점만 삭제할 수 있습니다.");
+        }
+        sections.remove(lastSection);
+    }
+
+    public List<Long> stationIds() {
+        if (sections.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        List<Long> stationIds = new ArrayList<>();
+        Section iter = firstSection();
+
+        while (!iter.equals(lastSection())) {
+            stationIds.add(iter.getUpStationId());
+            iter = findByUpStationId(iter.getDownStationId());
+        }
+
+        stationIds.add(iter.getUpStationId());
+        stationIds.add(iter.getDownStationId());
+        return stationIds;
+    }
+
+    private Section findByUpStationId(Long stationId) {
+        return sections.stream()
+                .filter(s -> s.matchUpStation(stationId))
+                .findAny()
+                .orElseThrow();
+    }
+
+    private Section firstSection() {
+        return sections.stream()
+                .filter(this::isFirstSection)
+                .findAny()
+                .orElseThrow();
+    }
+
+    private Section lastSection() {
+        return sections.stream()
+                .filter(this::isLastSection)
+                .findAny()
+                .orElseThrow();
+    }
+
+    private boolean isFirstSection(Section section) {
+        List<Long> downStationIds = sections.stream()
+                .map(Section::getDownStationId)
+                .collect(Collectors.toList());
+
+        return !downStationIds.contains(section.getUpStationId());
```

Comment:
이런 구현도 고려해볼 수 있을 것 같아요 
클린 코드의 한 구절입니다. 참고해주세요 😄 
> G29 : 부정 조건은 피하라
> 부정 조건은 긍정 조건보다 이해하기 어렵다. 가능하면 긍정 조건으로 표현한다.

```suggestion
return sections.stream()
                .map(Section::getDownStationId)
                .noneMatch(id -> id.equals(section.getUpStationId()));
```